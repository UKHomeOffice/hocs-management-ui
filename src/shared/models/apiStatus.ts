export default interface ApiStatus {
    display: string;
    timeoutPeriod: number;
    type: string;
}
