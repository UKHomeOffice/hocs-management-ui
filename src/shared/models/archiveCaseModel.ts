export default interface ArchiveCaseModel {
    caseReference: string;
    deleted: boolean;
}
