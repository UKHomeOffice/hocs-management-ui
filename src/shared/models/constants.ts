export const ADD_CHILD_TOPIC_ERROR_DESCRIPTION = 'Something went wrong while adding the child topic. Please try again.';
export const ADD_STANDARD_LINE_ERROR_DESCRIPTION = 'Something went wrong while adding the standard line. Please try again.';
export const ADD_TEMPLATE_ERROR_DESCRIPTION = 'Something went wrong while adding the template. Please try again.';
export const ADD_UNIT_ERROR_DESCRIPTION = 'Something went wrong while adding the unit. Please try again.';
export const ADD_USER_ERROR_DESCRIPTION = 'Something went wrong while adding the following users. Please try again.';
export const ADD_USER_ERROR_TITLE = 'There was an error adding the users';
export const DUPLICATE_CHILD_TOPIC_DESCRIPTION = 'A child topic with that name already exists';
export const DUPLICATE_UNIT_DESCRIPTION = 'A unit with those details already exists';
export const EMPTY_SUBMIT_ERROR_DESCRIPTION = 'Please select some users before submitting.';
export const EMPTY_SUBMIT_ERROR_TITLE = 'No users selected';
export const EMPTY_SUBMIT_TEAM_ERROR_DESCRIPTION = 'Please select a team before submitting.';
export const EMPTY_SUBMIT_TEAM_ERROR_TITLE = 'No team selected';
export const EMPTY_SUBMIT_DRAFT_QA_TEAM_ERROR_DESCRIPTION = 'Select a team for Initial Draft and QA response stages.';
export const EMPTY_SUBMIT_PRIVATE_MINISTER_TEAM_ERROR_DESCRIPTION = 'Select a team for Private Office and Minister sign off stages.';
export const EMPTY_SUBMIT_TOPIC_ERROR_DESCRIPTION = 'Select a topic before submitting.';
export const EMPTY_SUBMIT_TOPIC_ERROR_TITLE = 'No topic selected';
export const GENERAL_ERROR_TITLE = 'Something went wrong';
export const LOAD_PARENT_TOPICS_ERROR_DESCRIPTION = 'There was an error retrieving the parent topics.  Please try refreshing the page.';
export const LOAD_TOPICS_ERROR_DESCRIPTION = 'There was an error retrieving the parent topics.  Please try refreshing the page.';
export const LOAD_TOPIC_ERROR_DESCRIPTION = 'There was an error retrieving the parent topics.  Please try refreshing the page.';
export const LOAD_CASE_TYPES_DESCRIPTION = 'There was an error retrieving the case types.  Please try refreshing the page.';
export const LOAD_TEAM_ERROR_DESCRIPTION = 'There was an error retrieving the team.  Please try refreshing the page.';
export const LOAD_TEAM_MEMBERS_ERROR_DESCRIPTION = 'There was an error retrieving the teams members.  Please try refreshing the page.';
export const LOAD_TEAMS_ERROR_DESCRIPTION = 'There was an error retrieving the teams.  Please try refreshing the page.';
export const LOAD_USERS_ERROR_DESCRIPTION = 'There was an error retrieving the list of users.  Please try refreshing the page.';
export const REMOVE_FROM_TEAM_ALLOCATED_ERROR_DESCRIPTION = 'The user cannot be removed from the team as they have cases assigned';
export const REMOVE_FROM_TEAM_ERROR_DESCRIPTION = 'There was an error removing the user from the team';
export const VALIDATION_ERROR_TITLE = "Thereâ€™s a problem";
