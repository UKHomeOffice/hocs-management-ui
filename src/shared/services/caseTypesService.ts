import axios from 'axios';
import CaseType from '../models/caseType';
import Item from '../models/item';

const sortByLabel: ((a: Item, b: Item) => number) = (a, b) => {
    const labelA = a.label.toLocaleLowerCase();
    const labelB = b.label.toLocaleLowerCase();
    let comparison = 0;
    if (labelA > labelB) {
        comparison = 1;
    } else if (labelA < labelB) {
        comparison = -1;
    }
    return comparison;
};

export const getCaseTypes = () => new Promise<CaseType[]>((resolve, reject) =>
    axios.get('/api/case-types')
        .then((response) => {
            resolve(response.data.sort(sortByLabel));
        })
        .catch(error => reject(error))
);
