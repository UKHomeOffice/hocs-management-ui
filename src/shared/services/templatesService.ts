import axios from 'axios';

export const addTemplate = (template: FormData) => new Promise((resolve, reject) => axios
    .post('/api/templates', template)
    .then(() => resolve())
    .catch(reason => reject(reason))
);

export const deleteTemplate = (uuid: string) => new Promise((resolve, reject) => axios
    .delete(`/api/templates/${uuid}`)
    .then(() => resolve())
    .catch(reason => reject(reason))
);
