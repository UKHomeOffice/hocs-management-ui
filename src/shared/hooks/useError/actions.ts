import { FormError } from '../../models/formError';

export type AddFormError = {
    payload: FormError;
    type: 'AddFormError';
};

export type ClearError = {
    type: 'ClearError'
};

export type SetTitle = {
    payload: string;
    type: 'SetTitle';
};

export type SetDescription = {
    payload: string;
    type: 'SetDescription';
};

export type Action =
    AddFormError |
    ClearError |
    SetDescription |
    SetTitle;
