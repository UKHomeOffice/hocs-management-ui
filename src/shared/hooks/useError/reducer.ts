import { Action } from './actions';
import PageError from '../../models/pageError';

export const reducer = (state: PageError, action: Action): PageError => {
    switch (action.type) {
        case 'AddFormError':
            return {
                error: { ...state.error, formErrors: [...state.error ? state.error.formErrors || [] : [], action.payload] }
            };
        case 'ClearError':
            return { error: undefined };
        case 'SetDescription':
            return { error: { ...state.error, description: action.payload } };
        case 'SetTitle':
            return { error: { ...state.error, title: action.payload } };
    }
    return state;
};
