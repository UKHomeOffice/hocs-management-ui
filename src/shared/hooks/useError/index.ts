import React, { Reducer } from 'react';
import PageError from '../../models/pageError';
import { Action } from './actions';
import { reducer } from './reducer';
import { FormError } from '../../models/formError';
import ErrorMessage from 'shared/models/errorMessage';

const useError = (formErrorDescription?: string, formErrorTitle?: string)
    : [
        PageError,
        (formError: FormError) => void,
        () => void,
        (error: ErrorMessage) => void
    ] => {
    const [pageError, dispatch] = React.useReducer<Reducer<PageError, Action>>(reducer, {});

    const addFormError = (formError: FormError) => {
        (formErrorDescription || formErrorTitle) && setMessage({ title: formErrorTitle || '', description: formErrorDescription || '' });
        dispatch({ type: 'AddFormError', payload: formError });
    };
    const clearErrors = () => dispatch({ type: 'ClearError' });
    const setMessage = (errorMessage: ErrorMessage) => {
        dispatch({ type: 'SetDescription', payload: errorMessage.description });
        dispatch({ type: 'SetTitle', payload: errorMessage.title });
    };
    return [pageError, addFormError, clearErrors, setMessage];
};

export default useError;
