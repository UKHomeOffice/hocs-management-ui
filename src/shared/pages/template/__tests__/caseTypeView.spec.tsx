import React from 'react';
import { match, MemoryRouter } from 'react-router-dom';
import { createBrowserHistory, History, Location } from 'history';
import { act, wait, fireEvent, getByText, render, RenderResult } from '@testing-library/react';
import CaseTypeView from '../caseTypeView';
import * as CaseTypesService from '../../../services/caseTypesService';
import * as TemplatesService from '../../../services/templatesService';
import * as useError from '../../../hooks/useError';
import * as constants from '../../../models/constants';

let match: match<any>;
let history: History<any>;
let location: Location;

jest.mock('../../../services/caseTypesService', () => ({
    __esModule: true,
    getCaseType: jest.fn().mockReturnValue(Promise.resolve({
        displayName: '__displayName__',
        label: '__label__',
        shortCode: '__shortCode__',
        type: '__type__',
        value: '__caseType1__'
    })),
    getTemplatesForCaseType: jest.fn().mockReturnValue(Promise.resolve([{
        label: '__template1__',
        value: '__templateId1__'
    }, {
        label: '__template2__',
        value: '__templateId2__'
    }]))
}));

jest.mock('../../../services/templatesService', () => ({
    __esModule: true,
    deleteTemplate: jest.fn().mockReturnValue(Promise.resolve())
}));

const getCaseTypeSpy = jest.spyOn(CaseTypesService, 'getCaseType');
const getTemplatesForCaseTypeSpy = jest.spyOn(CaseTypesService, 'getTemplatesForCaseType');
const deleteTemplateSpy = jest.spyOn(TemplatesService, 'deleteTemplate');
const useErrorSpy = jest.spyOn(useError, 'default');
const setMessageSpy = jest.fn();
const clearErrorsSpy = jest.fn();

const renderComponent = () => render(
    <MemoryRouter>
        <CaseTypeView history={history} location={location} match={match}></CaseTypeView>
    </MemoryRouter>
);

beforeEach(() => {
    history = createBrowserHistory();
    match = {
        isExact: true,
        params: { type: '__type__' },
        path: '',
        url: ''
    };

    location = {
        hash: '',
        key: '',
        pathname: '',
        search: '',
        state: {}
    };
    useErrorSpy.mockImplementation(() => [{}, jest.fn(), clearErrorsSpy, setMessageSpy]);
    clearErrorsSpy.mockReset();
    setMessageSpy.mockReset();
});

describe('when the CaseTypeView component is mounted', () => {
    it('should render with default props', async () => {
        expect.assertions(3);
        let wrapper: RenderResult;
        act(() => {
            wrapper = renderComponent();
        });

        await wait(() => {
            expect(getCaseTypeSpy).toHaveBeenCalled();
            expect(getTemplatesForCaseTypeSpy).toHaveBeenCalled();
            expect(wrapper.container).toMatchSnapshot();
        });
    });
});

describe('when the Add template button is clicked', () => {
    it('should push a new page into the history', async () => {
        history.push = jest.fn();
        let wrapper: RenderResult;
        act(() => {
            wrapper = renderComponent();
        });

        await wait(async () => {
            const addTemplateButton = getByText(wrapper.container, 'Add template');
            fireEvent.click(addTemplateButton);
        });

        expect(history.push).toHaveBeenCalledWith('/case-type/__type__/add-template');
    });
});

describe('when the remove template button is clicked', () => {
    it('should remove the row from the templates table', async () => {
        let wrapper: RenderResult;
        act(() => {
            wrapper = renderComponent();
        });

        await wait(async () => {
            const selectedTemplate = getByText(wrapper.container, '__template1__');
            const row = (selectedTemplate.closest('tr'));
            const removeButton = getByText(row as HTMLElement, 'Remove');
            fireEvent.click(removeButton);
        });

        expect(deleteTemplateSpy).nthCalledWith(1, '__templateId1__');
        expect(getTemplatesForCaseTypeSpy).nthCalledWith(1, '__type__');

    });

    describe('and the service call fails', () => {
        beforeEach(() => {
            let wrapper: RenderResult;
            act(() => {
                wrapper = renderComponent();
            });

            wait(async () => {
                const selectedTemplate = getByText(wrapper.container, '__template1__');
                const row = (selectedTemplate.closest('tr'));
                const removeButton = getByText(row as HTMLElement, 'Remove');
                fireEvent.click(removeButton);
            });
        });

        describe('and its a 500 error', () => {
            beforeAll(() => {
                jest.spyOn(TemplatesService, 'deleteTemplate').mockImplementation(() => Promise.reject({ response: { status: 500 } }));
            });
            it('should set the error state', () => {
                wait(async () => {
                    expect(setMessageSpy).toHaveBeenCalledWith({ description: constants.REMOVE_TEMPLATE_ERROR_DESCRIPTION, title: constants.GENERAL_ERROR_TITLE });
                });
            });
            it('should call the clear errors method', () => {
                wait(async () => {
                    expect(clearErrorsSpy).toHaveBeenCalled();
                });
            });
        });
    });
});
