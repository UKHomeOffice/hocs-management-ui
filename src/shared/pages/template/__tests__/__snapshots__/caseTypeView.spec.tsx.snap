// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`when the CaseTypeView component is mounted should render with default props 1`] = `
<div>
  <div
    class="govuk-form-group"
  >
    <a
      class="govuk-back-link"
      href="/case-types"
    >
      Back
    </a>
    <div>
      <h1
        class="govuk-heading-xl"
      >
        View and remove templates
      </h1>
      <h2
        class="govuk-heading-l"
      />
      <div>
        <p
          class="govuk-body"
        >
          Loading...
        </p>
      </div>
      <button
        class="govuk-button govuk-!-margin-right-1 add-team-members-button"
        data-module="govuk-button"
        type="submit"
      >
        Add template
      </button>
    </div>
  </div>
</div>
`;
